fabtests/unit/mr_cache_evict.c

```diff
@@ -0,0 +1,685 @@
+	{
+	case SBRK:
+		return sbrk_alloc(size);
+	case MMAP:
+		return mmap_alloc(size);
+	case BRK:
+		return brk_alloc(size);
+	default:
+		return NULL;
+	}
+}
+
+static double timespec_diff_to_double(const struct timespec *start,
+				      const struct timespec *end)
+{
+	return (end->tv_sec - start->tv_sec ) +
+		(end->tv_nsec - start->tv_nsec ) / 1E9;
+}
+
+/* MR allocation function which returns the MR and the time in seconds to
+ * allocate the MR.
+ */
+static int mr_alloc(const void *buf, size_t buf_size, struct fid_mr **mr,
+		    double *alloc_time)
+{
+	uint64_t access = ft_info_to_mr_access(fi);
+	struct timespec start;
+	struct timespec end;
+	int ret;
+
+	ret = clock_gettime(CLOCK_MONOTONIC, &start);
+	if (ret) {
+		FT_UNIT_STRERR(err_buf, "clock_gettime failed", -errno);
+		return -errno;
+	}
+
+	ret = fi_mr_reg(domain, buf, buf_size, access, 0, FT_MR_KEY, 0, mr,
+			NULL);
+	if (ret != FI_SUCCESS) {
+		FT_UNIT_STRERR(err_buf, "fi_mr_reg failed", -errno);
+		return -errno;
+	}
+
+	ret = clock_gettime(CLOCK_MONOTONIC, &end);
+	if (ret) {
+		FT_UNIT_STRERR(err_buf, "clock_gettime failed", -errno);
+		return -errno;
+	}
+
+	*alloc_time = timespec_diff_to_double(&start, &end);
+
+	return 0;
+}
+
+/* Run a test verifing the eviction MR cache entries. The following is how the
+ * test works:
+ * 1. Prime CPU caches by allocating a priming MR. This MR is not used for
+ *    cache measurements.
+ * 2. Allocate a buffer using mem_alloc() with either MMAP, BRK, or SBRK. The
+ *    mem_alloc() allocator is constructed to return the same virtual address
+ *    during reallocation.
+ * 3. Measure MR allocation time of the mem_alloc() buffer. Since this buffer
+ *    does not exist in the cache, this should be a long allocation time.
+ * 4. Allocate another MR using the same mem_alloc() buffer. Since the buffer
+ *    has not been freed, a cached MR entry should be found. The cached MR
+ *    allocation time is measure. If that allocation time is not significantly
+ *    lower than the long/initial allocation time, it is assumed that the
+ *    provider does not support MR caching, and the test will exit.
+ * 5. If the provider supports caching, the mem_alloc() buffer is freed and
+ *    reallocated. Measures are inplace to have the reallocated mem_alloc()
+ *    buffer return the same virtual address. During this this time, the
+ *    provider's MR cache should experience an eviction.
+ * 6. A third MR is allocated using the mem_alloc() virtual address. Since the
+ *    provider should have evicted all entries in the cache using this virtual
+ *    address, the allocation time should be long again.
+ */
+static int mr_cache_test(enum mem_type type)
+{
+	void *prime_buf = NULL;
+	struct fid_mr *prime_mr = NULL;
+	void *buf = NULL;
+	struct fid_mr *mr = NULL;
+	double mr_alloc_time;
+	struct fid_mr *cached_mr = NULL;
+	double cached_mr_alloc_time;
+	struct fid_mr *realloc_mr = NULL;
+	double realloc_mr_alloc_time;
+	size_t buf_size = 4097;
```

Comment:
Can we test more buffer size instead of just 4097? BTW the number 4097 is interesting, maybe add a comment why this number is chosen?

Reply:
I think I'll make it an argument instead so users can try different values.